### DISCUSSION TITLE
QC for Adverse Event Assessment (AE)

### DISCUSSION THREAD

**Discussion by mattroumaya** [Gilead-Public/gsm#198] [General]
# QC for Adverse Event Assessment (AE)

- [ ] Documentation
    - [x] Function name captured in [Roxygen title](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block)(e.g. "Adverse Event Assessment")
    - [x] Assessment Purpose captured in [Roxygen description](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."
    - [ ] Input Data Requirements are described in the the [Roxygen details](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) under a Data section (`#' @details # Data Requirements`).
    - [x] Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the [Roxygen details](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) under a Statistics section (`#' @details # Statistical Assumptions`).
    - [ ] Parameters have detailed documentation using `@param`. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)
    - [x] All external dependencies are captured. Use `@importFrom` when importing 3 or fewer functions, and use `@import` otherwise. 
    - [x] Output Data Standards are described under `@returns`
    - [x] At least 1 example is provided under `@examples`
- [ ] Error Checking
    - [ ]  Basic checks for all all parameters should be included using `stopifnot()` or similar logic. (e.g. `stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput))`)
    - [x] Tests confirm that `stopifnot()` parameter checks are working as expected. 
    - [x] Tests confirm that the input data has required columns (if any)
    - [x] Tests confirm that the output data has the expected columns
    - [ ] Tests confirm intended functionality for each parameter
    - [x] Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed. 
- [ ] Basic QC
    - [ ] Assessment has User Requirements + Qualification tests captured using {valtools} framework. Report is generating as expected and all checks are passing. 
    - [x] Code is well commented and easy to read
    - [x] No file paths or other company-specific data is present
    - [x] Function calls from non-tidyverse dependencies are called via `::`
    - [x] devtools::check() passes with no errors/warnings/notes
    - [x] package docs up to date. devtools::document() doesn't change any files
    - [x] Uses tidyverse best practices for standard data manipulation (start discussion thread if unclear)
    - [x] All new dependencies add significant value (start discussion thread)
    - [x] All GitHub Actions run with no errors

_Originally posted by @mattroumaya in https://github.com/Gilead-BioStats/gsm/pull/194#issuecomment-1051244379_

**Comment by mattroumaya** [comment]
**Input Data Requirements are described in the Roxygen details under a Data section (#' @details # Data Requirements).**
- [AE_Assess is missing a description for `Rate`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/AE_Assess.R#L12-L15)
  - Suggested fix: Add - `Rate` - Rate of exposure

**Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
- [AE_Map_Raw should note required columns for `dfAE`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/AE_Map_Raw.R#L23)
  - Suggested fix: _AE dataset with required column SUBJID, and rows for each AE record_

**Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
- [AE_Assess](https://github.com/Gilead-BioStats/gsm/blob/dev/R/AE_Assess.R#L45-L51):
  - Missing error handling for `vThreshold`. Suggest adding:
```r
"vThreshold is not numeric" = is.numeric(vThreshold),
"vThreshold must be length 2" = length(vThrehsold) == 2
```
- Stricter error handling for `cMethod`. (e.g., this currently passes: `AE_Assess(dfInput, cMethod = c("wilcoxon", "poisson")`
- Suggest adding:
```r
"cMethod must be length 1" = length(cMethod) == 1
```

**Tests confirm that stopifnot() parameter checks are working as expected.**
- AE_Assess: [these lines](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_AE_Assess.R#L19-L34) can be combined into one test

**Tests confirm intended functionality for each parameter**
- AE_Assess
  - Tests do not explicitly check `poisson` or `wilcoxon` arguments to `cMethod`
  - Tests do not check specific `vThreshold` values (though not sure if this is needed)

- AE_Map_Raw
  - Tests do not check `strExposureCol` parameter

**Assessment has User Requirements + Qualification tests captured using {valtools} framework. Report is generating as expected and all checks are passing.**
- Qualification needs to be finalized (double-programming; add `AE_Map_Adam`)

**Reply by jwildfire** [reply]
> **Input Data Requirements are described in the Roxygen details under a Data section (#' @details # Data Requirements).**
> 
> * [AE_Assess is missing a description for `Rate`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/AE_Assess.R#L12-L15)
>   
>   * Suggested fix: Add - `Rate` - Rate of exposure 

Agreed. 

> **Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
> 
> * [AE_Map_Raw should note required columns for `dfAE`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/AE_Map_Raw.R#L23)
>   
>   * Suggested fix: _AE dataset with required column SUBJID, and rows for each AE record_

👍 

> **Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
> 
> * [AE_Assess](https://github.com/Gilead-BioStats/gsm/blob/dev/R/AE_Assess.R#L45-L51):
>   
>   * Missing error handling for `vThreshold`. Suggest adding:
> 
> ```r
> "vThreshold is not numeric" = is.numeric(vThreshold),
> "vThreshold must be length 2" = length(vThrehsold) == 2
> ```

No update needed. stopifnot is specified for each method. 

> * Stricter error handling for `cMethod`. (e.g., this currently passes: `AE_Assess(dfInput, cMethod = c("wilcoxon", "poisson")`
> * Suggest adding:
> 
> ```r
> "cMethod must be length 1" = length(cMethod) == 1
> ```

👍 
 
> **Tests confirm that stopifnot() parameter checks are working as expected.**
> 
> * AE_Assess: [these lines](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_AE_Assess.R#L19-L34) can be combined into one test

👍 

> **Tests confirm intended functionality for each parameter**
> 
> * AE_Assess
>   
>   * Tests do not explicitly check `poisson` or `wilcoxon` arguments to `cMethod`

Lets add a basic test to confirm that valid methods return data in the correct structure (e.g. 1 rec per site with expected columns)

>   * Tests do not check specific `vThreshold` values (though not sure if this is needed)

Out of scope `Assess` covered in `flag()`

> * AE_Map_Raw
>   
>   * Tests do not check `strExposureCol` parameter

👍  Add a quick test for strExporureCol = "TimeOnTreatment" 


> **Assessment has User Requirements + Qualification tests captured using {valtools} framework. Report is generating as expected and all checks are passing.**
> 
> * Qualification needs to be finalized (double-programming; add `AE_Map_Adam`)

👍

**Reply by jwildfire** [reply]
Move forward with v0.1. File 2 issues for v0.2. One for Assess and one for Map Raw. Link to this discussion in the PR for reference.
